indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java

```diff
@@ -79,36 +80,46 @@ public HandlingInputRowIterator build()
   }
 
   private static HandlingInputRowIterator.InputRowHandler createOnlySingleDimensionValueRowsHandler(
-      String partitionDimension
+      List<String> partitionDimensions
   )
   {
-    return inputRow -> {
-      int dimensionValueCount = getSingleOrNullDimensionValueCount(inputRow, partitionDimension);
-      return dimensionValueCount != 1;
-    };
+    return inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> dimValueCount != 1);
   }
 
   private static HandlingInputRowIterator.InputRowHandler createOnlySingleOrNullDimensionValueRowsHandler(
-      String partitionDimension
+      List<String> partitionDimensions
   )
   {
-    return inputRow -> {
-      int dimensionValueCount = getSingleOrNullDimensionValueCount(inputRow, partitionDimension);
-      return dimensionValueCount > 1;  // Rows.objectToStrings() returns an empty list for a single null value
-    };
+    // Rows.objectToStrings() returns an empty list for a single null value
+    return inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> dimValueCount > 1);
   }
 
-  private static int getSingleOrNullDimensionValueCount(InputRow inputRow, String partitionDimension)
+  /**
+   * @param valueCountPredicate Predicate that must be satisfied
+   *                            for atleast one of the partitionDimensions for the row to be marked as handled.
+   * @return true when the given InputRow should be marked handled
+   * and need not be processed further.
+   */
+  private static boolean isRowHandled(
+      InputRow inputRow,
+      List<String> partitionDimensions,
+      Predicate<Integer> valueCountPredicate
+  )
   {
-    List<String> dimensionValues = inputRow.getDimension(partitionDimension);
-    int dimensionValueCount = dimensionValues.size();
-    if (dimensionValueCount > 1) {
-      throw new IAE(
-          "Cannot partition on multi-value dimension [%s] for input row [%s]",
-          partitionDimension,
-          inputRow
-      );
+    for (String dimension : partitionDimensions) {
+      int dimensionValueCount = inputRow.getDimension(dimension).size();
+      if (dimensionValueCount > 1) {
+        throw new IAE(
+            "Cannot partition on multi-value dimension [%s] for input row [%s]",
+            dimension,
+            inputRow
+        );
+      } else if (valueCountPredicate.test(dimensionValueCount)) {
```

Comment:
so the `valueCountPredicate` is `dimValueCount -> dimValueCount > 1` but if dimensionValueCount is greater than 1, we will throw an exception. so this should always return false. isn't it?

Reply:
ok. there is one more value predicate. thinking more on it, `inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> dimValueCount > 1)` can very well be inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> false)`. Just trying to understand this logic better.

Reply:
Refactored this logic to have two methods:
- `ensureNoMultiValueDimensions()` which checks for multiple values and throws IAE
- `hasEmptyDimensions()`